lib/queue.js

```diff
@@ -25,276 +48,197 @@ var FetchQueue = function() {
 module.exports = FetchQueue;
 
 FetchQueue.prototype = [];
-FetchQueue.prototype.add = function(protocol, domain, port, path, depth, callback) {
+FetchQueue.prototype.add = function(queueItem, force, callback) {
+    callback = callback instanceof Function ? callback : noop;
+    var queue = this;
 
-    // For legacy reasons
-    if (depth instanceof Function) {
-        callback = depth;
-        depth = 1;
+    if (arguments.length === 2) {
+        callback = force;
+        force = undefined;
     }
 
-    depth = depth || 1;
-    callback = callback && callback instanceof Function ? callback : function() {};
-    var self = this;
-
-    // Ensure all variables conform to reasonable defaults
-    protocol = protocol === "https" ? "https" : "http";
-
-    if (isNaN(port) || !port) {
-        return callback(new Error("Port must be numeric!"));
+    function addToQueue() {
+        queue.scanIndex[queueItem.url] = true;
+        queueItem.status = "queued";
+        queue.push(queueItem);
+        callback(null, queueItem);
     }
 
-    var url = protocol + "://" + domain + (port !== 80 ? ":" + port : "") + path;
-
-    self.exists(protocol, domain, port, path,
-        function(err, exists) {
-            if (err) {
-                return callback(err);
-            }
-
-            if (!exists) {
-                var queueItem = {
-                    url: url,
-                    protocol: protocol,
-                    host: domain,
-                    port: port,
-                    path: path,
-                    depth: depth,
-                    fetched: false,
-                    status: "queued",
-                    stateData: {}
-                };
-
-                self.push(queueItem);
-                callback(null, queueItem);
+    queue.exists(queueItem.url, function(err, exists) {
+        if (err) {
+            callback(err);
+        } else if (!exists) {
+            addToQueue();
+        } else if (force) {
+            if (queue.indexOf(queueItem) > -1) {
```

Comment:
Good catch — at least with the current queue implementation, not having this condition would mean that statuses could be updated for more than one queue item simultaneously, completely erroneously.